### ISSUE TITLE
Data binding not working for checkboxes

### ISSUE THREAD

**Issue by AdrianBav** [calebporzio/livewire#173]
**Describe the bug**
Data binding does not appear to be working for checkboxes. Other input types, including radio buttons do work as expected. 

**To Reproduce**
1. Source of App\Http\Livewire\Demo.php:
```
<?php

namespace App\Http\Livewire;

use Livewire\Component;

class Demo extends Component
{
    public $color = 'none';
    public $cars = [];

    public function render()
    {
        return view('livewire.demo');
    }
}
```

2. Source of resource\view\livewire\demo.blade.php:
```
<div>
    <form>
        <fieldset>
            <legend>Colors</legend>
            <div>
                <label for="color">None</label>
                <input type="radio" wire:model="color" name="color" value="none">
            </div>
            <div>
                <label for="color">Red</label>
                <input type="radio" wire:model="color" name="color" value="red">
            </div>
            <div>
                <label for="color">Blue</label>
                <input type="radio" wire:model="color" name="color" value="blue">
            </div>
        </fieldset>

        <fieldset>
            <legend>Cars</legend>
            <div>
                <label for="cars">Honda</label>
                <input type="checkbox" wire:model="cars" name="cars" value="honda">
            </div>
            <div>
                <label for="cars">Ford</label>
                <input type="checkbox" wire:model="cars" name="cars" value="ford">
            </div>
            <div>
                <label for="cars">Golf</label>
                <input type="checkbox" wire:model="cars" name="cars" value="golf">
            </div>
        </fieldset>
    </form>

    <hr>

    <div>
        <p>The color is: {{ $color }}</p>

        <p>Selected cars:</p>
        <ul>
            @forelse ($cars as $car)
            <li>{{ $car }}</li>
            @empty
            <li>No cars selected</li>
            @endforelse
        </ul>
    </div>
</div>
```

**Expected behavior**
On setting the initial state of the model to an empty array, I would expect all checkboxes bound to the model to be unchecked. On checking a checkbox, I would expect the value of that checkbox to be sent to the server. 

**Desktop (please complete the following information):**
 - Browser, Chrome, Windows

**Comment by calebporzio** [comment]
You are right @AdrianBav, it appears checkboxes are not handled properly in Livewire.

There are two things that need to be accounted for:
* When a `value` attribute is present on the `<input>` element when it is checked, the Livewire property should update to that value, not a boolean like it currently does.
* When multiple checkboxes with the same `wire:model` value and different `value` attribute values are clicked, the Livewire property should be bound as an array (which is what you are describing)

This is not a trivial update, but of course, is necessary. Thanks for pointing it out.

**Comment by AdrianBav** [comment]
Thanks Caleb. As Caleb rightly says, single checkboxes are working correctly. As a temporary fix I used nested data binding to achieve something similar:

demo.blade.php:
```
@foreach ($cars as $car)
<div>
    <label for="cars">{{ $car }}</label>
    <input type="checkbox" wire:model="cars.{{ $car }}" name="cars" value="1">
</div>
@endforeach
```

**Comment by calebporzio** [comment]
Just checked what Vue does, and it sets it to a boolean regardless of the presence of a `value` attribute

**Comment by intrepidws** [comment]
@calebporzio Before I dive in here, please keep in mind that I've used Vue maybe twice in my life and I could very well be wrong here. That said, my reading of the below link is that while Vue essentially does a boolean (true/false or checked/unchecked), it does allow you to specify attributes for `true-value` and `false-value`.

https://vuejs.org/v2/guide/forms.html#Value-Bindings

It's very possible that I've just been using the HTML incorrectly, but I find it very useful to be able to specify a value for a checkbox. My use case is that I am making a registration form for an event and I need to make a bunch of checkboxes for optional events/sessions, etc. Before using Livewire, I would do something like this:

```
<input type="checkbox" name="events[]" value="123" /> Event 123
<input type="checkbox" name="events[]" value="456" /> Event 456
```

Since Livewire doesn't allow an empty array like that, I've been doing the following:

```
<input type="checkbox" name="events[123]" value="123" /> Event 123
<input type="checkbox" name="events[456]" value="456" /> Event 456
```

While I could certainly work with that if need be, it would prevent me from easily doing something like the following Validation to ensure that the codes exist in the database.

```
$validatedData = $this->validate([
    'events.*' => ['required', 'exists:code,id'],
]);
```

I guess my point is that it's much easier to work with the "value", rather than the "key" of the resulting array. Anyway, I'm happy to open another issue for this if you think it's worthy of some further discussion - just let me know.

**Comment by calebporzio** [comment]
@intrepidws - I believe what you're looking for is the current functionality:

* Behavior
![Sep-04-2019 10-02-25](https://user-images.githubusercontent.com/3670578/64261882-3dc5d800-cefb-11e9-9b99-d6f9614e3fca.gif)

* Code
```
... Component {
    public $event = [];
...
}

<div>
    <input type="checkbox" wire:model="event", value="123">
    <input type="checkbox" wire:model="event", value="456">
    <div>@json($event)</div>
</div>
```

**Comment by ryanscherler** [comment]
I think the current implementation doesn't allow for the model to be initially checked. In the example above, if you were to already fill the model with:

```
$this->event = ["123"];
```

Then the first checkbox would be checked (which is good), but for some reason now cannot be 'unchecked'?

**Comment by redbastie** [comment]
I ran into the same issue, and resolved it by appending the blade `$loop` index:

                        @foreach($field->optionsArray() as $option)
                            <label class="btn btn-lg btn-primary">
                                <input type="{{ $field->type == 'sbutton' ? 'radio' : 'checkbox' }}" value="{{ $option }}"
                                       wire:model="{{ $field->dataName() }}{{ $field->type == 'mbutton' ? '.' . $loop->index : null }}">
                                {{ $option }}
                            </label>
                        @endforeach

**Comment by wrabit** [comment]
Similar to @redbastie but am using the id's of the values that I will save:

                    @foreach($items as $item)
                        <input type="checkbox" value="1"
                                   wire:model="myItems.{{ $item->id }}">
                            {{ $item->name }}
                    @endforeach

And then in my save method in the component I use `array_keys` on `$myItems`.

To initialise the checked boxes, in `mount()` I create `myItems` like so:

`$this->myItems = array_fill_keys($user->items->toArray(), 1);`

**Comment by rafael1delgado** [comment]
Hi, it's working for me right now.

Thanks.

PD: It took 2 days to find it would work

**Comment by adeayobandung** [comment]
> @intrepidws - I believe what you're looking for is the current functionality:
> 
> * Behavior
>   ![Sep-04-2019 10-02-25](https://user-images.githubusercontent.com/3670578/64261882-3dc5d800-cefb-11e9-9b99-d6f9614e3fca.gif)
> * Code
> 
> ```
> ... Component {
>     public $event = [];
> ...
> }
> 
> <div>
>     <input type="checkbox" wire:model="event", value="123">
>     <input type="checkbox" wire:model="event", value="456">
>     <div>@json($event)</div>
> </div>
> ```

This is good code

**Comment by snapey** [comment]
In order to prime the checkboxes from my model, I found it necessary to convert the IDs to strings

mount;
```
        $this->groups = $user->groups()->pluck('id')->map(
                                function($group) {
                                    return strval($group);
                                });
```

**Comment by tdanandeh** [comment]
> public $event = [];

good answer...thank you

**Comment by aurelien31** [comment]
> In order to prime the checkboxes from my model, I found it necessary to convert the IDs to strings
> 
> mount;
> 
> ```
>         $this->groups = $user->groups()->pluck('id')->map(
>                                 function($group) {
>                                     return strval($group);
>                                 });
> ```
> 
> or
> 
> ```
>         $this->groups = $user->groups()->pluck('id')->map(fn($g) => strval($g));
> ```

Same problem here, Livewire seems to update "numbers" checboxes value (like ids) to string.

After many search and try, thanks @snapey for the more simplest solution I've found.

**Comment by bhulsman** [comment]
Using...

```
public array $rows = [];
```

...and...

```
@foreach($rows as $row)
    <div wire:key="{{ $row->id }}">
        <input wire:model.defer="rows.{{ $row->id }}" type="checkbox" />
    </div>
@endforeach

<button wire:click="doSomethingWithRows">Submit</button>
```

Say you have 4 options and click 3 of them, the first click on 'submit' shows those 3, thats correct. But when you also enable the 4rd and then click the 'submit' button, only the 4rd shows up, while all are 'checked'.

What am I missing?

**Comment by joshhanley** [comment]
@bhulsman you are looping through `$rows` and calling object properties on them like `$row->id` but also binding to `$rows`, which will change the values to true/false.

Try changing it so you have your `$rows` is only your collection of models, and `$selectedRows` is what you bind to with `wire:model`.

```php
public EloquentCollection $rows;
public array $selectedRows;
```

```blade
@foreach($rows as $row)
    <div wire:key="{{ $row->id }}">
        <input wire:model.defer="selectedRows.{{ $row->id }}" type="checkbox" />
    </div>
@endforeach

<button wire:click="doSomethingWithRows">Submit</button>
```

Hope this helps!

**Comment by bhulsman** [comment]
> @bhulsman you are looping through `$rows` and calling object properties on them like `$row->id` but also binding to `$rows`, which will change the values to true/false.
> 
> Try changing it so you have your `$rows` is only your collection of models, and `$selectedRows` is what you bind to with `wire:model`.
> 
> ```
> public EloquentCollection $rows;
> public array $selectedRows;
> ```
> 
> ```
> @foreach($rows as $row)
>     <div wire:key="{{ $row->id }}">
>         <input wire:model.defer="selectedRows.{{ $row->id }}" type="checkbox" />
>     </div>
> @endforeach
> 
> <button wire:click="doSomethingWithRows">Submit</button>
> ```
> 
> Hope this helps!

Thanks, I did not show this correctly but it's `rowsToCopy` in my code, my bad!

**Comment by bhulsman** [comment]
The solution is to prefill the array apparently.

**Comment by code23-isaac** [comment]
Any way to use the debounce method for multiple checkboxes bound to the same model? I'm finding that debouncing only applies to each individual input, rather than all of the checkboxes.

**Comment by snapey** [comment]
@code23-isaac How can you debounce a checkbox?  Its checked or it isn't?

**Comment by code23-isaac** [comment]
I achieved what i'm after by x-modelling to an alpine prop and manually debouncing $wire.set with a timeout.

**Comment by tigusigalpa** [comment]
> Any way to use the debounce method for multiple checkboxes bound to the same model? I'm finding that debouncing only applies to each individual input, rather than all of the checkboxes.

Use radiobutton instead checkbox

**Comment by alesf** [comment]
> @intrepidws - I believe what you're looking for is the current functionality:
> 
> * Behavior
>   ![Sep-04-2019 10-02-25](https://user-images.githubusercontent.com/3670578/64261882-3dc5d800-cefb-11e9-9b99-d6f9614e3fca.gif)
>         
>           [
>             
>               ![Sep-04-2019 10-02-25](https://user-images.githubusercontent.com/3670578/64261882-3dc5d800-cefb-11e9-9b99-d6f9614e3fca.gif)
>             
>           ](https://user-images.githubusercontent.com/3670578/64261882-3dc5d800-cefb-11e9-9b99-d6f9614e3fca.gif)
>           
>           
>             
>               
>                 
>               
>               
>                 
>                 
>               
>             
>             [
>               
>                 
>               
>             ](https://user-images.githubusercontent.com/3670578/64261882-3dc5d800-cefb-11e9-9b99-d6f9614e3fca.gif)
> * Code
> 
> ```
> ... Component {
>     public $event = [];
> ...
> }
> 
> <div>
>     <input type="checkbox" wire:model="event", value="123">
>     <input type="checkbox" wire:model="event", value="456">
>     <div>@json($event)</div>
> </div>
> ```

This does not work with nested arrays:

Example:

```
... Component {
    public $event = [
	'nested' => [],
    ];
...
}

<div>
    <input type="checkbox" wire:model="event.nested", value="123">
    <input type="checkbox" wire:model="event.nested", value="456">
    <div>@json($event['nested'])</div>
</div>
```

@calebporzio is there any way to implement it?

**Comment by code23-isaac** [comment]
@alesf is $event a reserved name? have you tried naming it something else?

**Comment by alesf** [comment]
@code23-isaac
This is a weird one.
I was banging my head with this for a couple of hours, but in the end figured it is not implemented the way this would work.

After your suggestion to check for reserved name I changed the variable name and it was working.
After setting the default (empty) values in `mounted` method it's now also working for original variable.
So I guess it was actually due to some Livewire caching or something.

**Comment by burfbari** [comment]
> When a value attribute is present on the <input> element when it is checked, the Livewire property should update to that value, not a boolean like it currently does.

as I have got multiple Livewire components (each having 1 checkbox)
`<livewire:components.toggle-admin :userID="$user->userID" :wire:key="'user-'.$user->userID">`

the `id` seems to be getting passed fine
```
public $userID;
```

but value seems to be messed up:
```
<div x-data="{checked : '{{ $checked }}'}">
{{ $userID }}
<input
        class="hidden appearance-none"
        type="checkbox"
        id="check"
        value="{{ $userID }}"
        wire:model="list"
        x-model="checked"
        @click="$wire.toggleAdminRights('{{ $userID }}')"
    >
```

The `userID` renders fine but when setting this as `value`, there seems to an issue.

![livewire-checkboxes](https://github.com/livewire/livewire/assets/52349056/e3ae457b-4f57-4316-96d2-c0f3bca77b31)

Any ideas?

**Comment by code23-isaac** [comment]
@burfbari not sure why you're modelling the checkbox with both livewire and alpine, but you should probably make your own  post as this is getting sidetracked

**Comment by kirkbushell** [comment]
Is this broken again in 3? I'm seeing some real weirdness in Livewire 3:


https://github.com/livewire/livewire/assets/65171/dd41acdc-77b6-4b00-a9f9-57528737011b


This is the code:

```php
<x-forms.checkbox value="email" wire:model.live="form.notifications_methods">Email</x-forms.checkbox>
<x-forms.checkbox value="push" wire:model.live="form.notifications_methods">Push (Browser)</x-forms.checkbox>
```

The component is just a wrap for the checkbox element, with some default classes applied.

This is the form class:

```php
class ProfileForm extends Form
{
	public $notifications_methods = [];
}
```

**Comment by snapey** [comment]
try naming the checkboxes with array syntax

    name="notifications_methods[]"

**Comment by kirkbushell** [comment]
> try naming the checkboxes with array syntax
> 
> ```
> name="notifications_methods[]"
> ```
> 
> but if you only have two values, just create individual public properties for them

1. This is not what is advised on livewire docs.
2. That's really not desirable - the amount of code required to manage the structures is a PITA

=\

**Comment by joshhanley** [comment]
@kirkbushell yeah what you are after should work. Any chance you could submit a failing test PR? Thanks!

**Comment by snapey** [comment]
@kirkbushell The behaviour is as if the `value` is not being passed through to the checkbox component.

**Comment by kirkbushell** [comment]
> @kirkbushell The behaviour is as if the `value` is not being passed through to the checkbox component
> 
> Check the generated html?

This is from my blade component:

```html
<input type="checkbox" {{ $attributes->merge(['class' => 'bg-primary-100 text-secondary-600 focus:ring-primary-400']) }} />
```

Ie. It was passing through the value.

This was the actual rendered HTML:

```html
<input type="checkbox" class="bg-primary-100 text-secondary-600 focus:ring-primary-400" value="push" wire:model.live="form.notifications_method">
```

Livewire (or alpine) appears to be bugging out, here. I'm not entirely sure which, but multi-checkbox simply does not work.

Note that it doesn't matter if it's a multi-checkbox or not - with checkboxes, it only appears to be passing true as a value.

**Comment by kirkbushell** [comment]
> @kirkbushell yeah what you are after should work. Any chance you could submit a failing test PR? Thanks!

Apologies, where can I find details on that?

**Comment by joshhanley** [comment]
@kirkbushell Have a look at our contribution guide https://livewire.laravel.com/docs/contribution-guide

If you have troubles actually running tests locally (sometimes if people use docker then disk can be hard to get working), just write one anyway and push up a draft PR with it and we can help get it going. 🙂

**Comment by kirkbushell** [comment]
@joshhanley 

Here you are - this discussion raised in December last year highlights the issue:

https://github.com/livewire/livewire/discussions/7581

This is exactly what I was experiencing as well, and even more weirdness.

**Comment by joshhanley** [comment]
@kirkbushell Oh I believed you (I hope it didn't come across as otherwise), that's why I asked for a failing test so we can look into what's going on 🙂

**Comment by kirkbushell** [comment]
> @kirkbushell Oh I believed you (I hope it didn't come across as otherwise), that's why I asked for a failing test so we can look into what's going on 🙂

Oh, yeah all good - but provided that discussion above, as it shows what's going on, as they provided a working test.

**Comment by snapey** [comment]
I created a wirebox with a few ideas https://wirebox.app/b/xeg6g (which showed the described behaviour only when no value)

and then discovered the wirebox in #7581 and could not see any problem with its behaviour

and then created a fork which more clearly illustrated the lack of any issue.

**Comment by joshhanley** [comment]
Thanks @snapey, @kirkbushell any chance you could update to the latest version of Livewire v3.4.4 and test your set up again? If it still doesn't work, can you submit a PR containing a failing Dusk test and so we can look into it? Thanks!
